src/lib/logical_query_plan/abstract_lqp_node.hpp

```diff
@@ -172,9 +172,23 @@ class AbstractLQPNode : public std::enable_shared_from_this<AbstractLQPNode> {
 
   /**
    * @return The functional dependencies valid for this node. See functional_dependency.hpp for documentation.
-   *         By default, functional dependencies from both sides are forwarded. Nodes may override this behavior.
+   *         They are collected as follows:
+   *          (1) FDs are derived from the current node's unique constraints.
+   *          (2) FDs are provided by previous nodes by passing them up the LQP tree.
    */
-  virtual std::vector<FunctionalDependency> functional_dependencies() const;
+  std::vector<FunctionalDependency> functional_dependencies() const;
+
+  /**
+   * Helper function that passes FDs to upper nodes. By default, it passes FDs from the left input node.
+   * Some nodes may override this function to
```

Comment:
Have you considered a non-virtual method `_fds_from_unique_constraints`, which is then called by `virtual functional_dependencies`? For some reason, that feels much more intuitive to me.

Reply:
Not really. It could be an improvement code-wise. But, I do not see a big architectural change here.

Calling `_fds_from_unique_constraints` over and over again from within the LQP tree means a lot of overhead. It only makes sense when we lose unique constraints (e.g. JoinNode). Otherwise, it should be called once as the last step before returning to the user.

Reply:
Could you describe where this would be an overhead in more detail? Good chance that I'm missing something here

Reply:
Think of an LQP as follows: 
```
  A
  |
  B
  |
  C (JoinNode)
 /  \
D    E
```
**Naive implementation**
Calling `functional_dependencies()` on A leads to:
* Call `unique_constraints()` on A
  * Call `unique_constraints()` on B
    * Call `unique_constraints()` on C
      * Call `unique_constraints()` on D
      * Call `unique_constraints()` on E
* Call `functional_dependencies()` on B
  * Call `unique_constraints()` on B
    * Call `unique_constraints()` on C
      * Call `unique_constraints()` on D
      * Call `unique_constraints()` on E
* Call `functional_dependencies()` on C
    * Call `unique_constraints()` on C
      * Call `unique_constraints()` on D
      * Call `unique_constraints()` on E
* Call `functional_dependencies()` on D
  * Call `unique_constraints()` on D
* Call `functional_dependencies()` on E
  * Call `unique_constraints()` on E

**Optimized implementation**
Calling `functional_dependencies()` on A leads to:
* Call `unique_constraints()` on A
  * Call `unique_constraints()` on B
    * Call `unique_constraints()` on C
      * Call `unique_constraints()` on D
      * Call `unique_constraints()` on E
* Call `pass_functional_dependencies()` on B
* Call `pass_functional_dependencies()` on C (JoinNode)
  * Call `functional_dependencies()` on D
    * Call `unique_constraints()` on D
  * Call `functional_dependencies()` on E
    * Call `unique_constraints()` on E


Hope it helps. Regarding the special Join-handling, see https://github.com/hyrise/hyrise/pull/2189#discussion_r464344365

Reply:
Yepp, that makes sense. Good call. I don't like the method's name too much though. Maybe `_non_unique_functional_dependencies` instead of `pass_functional_dependencies`?

Reply:
Generally, I don't think that "pass" is the best word here, neither in the method name nor in the comment.

Reply:
I already had a hard time thinking about a good, short name for this type of helper function. 

To summarize: The function returns FDs that cannot be derived from unique constraints in upper nodes. 
* Some FDs might have become invalid due to projection or nullability changes
* No duplicate FDs in the output

To be honest, I do not like `_non_unique_functional_dependencies` either. What is a non-unique FD? :/

Reply:
Maybe `non_derivable_functional_dependencies()` ?

Reply:
non-trivial?

Reply:
"non-trivial" works great for me 👍